### ISSUE TITLE
Unable to run \Magento\Downloadable\Test\Unit\Helper\DownloadTest without internet connection / dns resolution

### ISSUE THREAD

**Issue by adrian-martinez-interactiv4** [magento/magento2#23521]
### Preconditions (*)
1. Magento 2.3.2, just created project, no setup:install
2. PHP 7.2.19

### Steps to reproduce (*)

1. Create Magento 2.3.2 project
2. Run unit tests (at least \Magento\Downloadable\Test\Unit\Helper\DownloadTest) without internet connection

### Expected result (*)

1. Test should pass (image below shows test execution with internet connection): 

<img width="1377" alt="Captura de pantalla 2019-07-01 a las 13 04 04" src="https://user-images.githubusercontent.com/17545750/60431731-c0cb6900-9c00-11e9-8396-4f26933f7c80.png">

### Actual result (*)

1. Test fails with message  (image below shows test execution without internet connection): `PHPUnit\Framework\Exception: Warning: get_headers(): php_network_getaddresses: getaddrinfo failed: nodename nor servname provided, or not known in /Users/adrianmartinez/Sites/2.3/magento/vendor/magento/module-downloadable/Helper/Download.php:264.`

<img width="1380" alt="Captura de pantalla 2019-07-01 a las 13 04 58" src="https://user-images.githubusercontent.com/17545750/60431814-f6705200-9c00-11e9-8e33-4eda345cb3bd.png">

### Extended description

This behaviour cannot be reproduced in 2.2.8, test passes even if there is no internet connection. Comparing code both versions, I've seen this snippet of code has been added in new version of tested class `\Magento\Downloadable\Helper\Download`:
```
    public function setResource($resourceFile, $linkType = self::LINK_TYPE_FILE)
    {
        (...)
        
        /**
        * check header for urls
        */
        if ($linkType === self::LINK_TYPE_URL) {
            $headers = array_change_key_case(get_headers($this->_resourceFile, 1), CASE_LOWER);
            if (isset($headers['location'])) {
                $this->_resourceFile  = is_array($headers['location']) ? current($headers['location'])
                    : $headers['location'];
            }
        }
        
        (...)
    }
```

This new code looks quite risky and not safe, I mean, even if problem is the test not passing, this piece of code has some assumptions that can break execution at any time.

First of all, test relies on a test url: `\Magento\Downloadable\Test\Unit\Helper\DownloadTest::URL = 'http://example.com'`, that may exist or not, or could be down at any time. This means this test could even fail with internet connection if that site is down or DNS servers are not available.

In second place, it tries to extract headers using `get_headers` php built-in function, and passes the result directly to `array_change_key_case`, without any kind of check. `get_headers` may emit a warning if it is not able to get header of supplied url, but even if silenced with `@get_headers`, execution throws another warning due to `get_headers` returning false instead of an array, and `array_change_key_case` complaining about receiving a boolean parameter instead of an array:

```
1) Magento\Downloadable\Test\Unit\Helper\DownloadTest::testGetFileSizeUrl
PHPUnit\Framework\Exception: Warning: array_change_key_case() expects parameter 1 to be array, boolean given in /Users/adrianmartinez/Sites/2.3/magento/vendor/magento/module-downloadable/Helper/Download.php:264.
```

I think that piece of code should be refactored through a HeadersResolver (or another more appropriate name), responsible of returning always a consistent return value when requested for some headers (always return an array), and use this new class in \Magento\Downloadable\Helper\Download, so it can be mocked for testing.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Charlie 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-17920` were created

**Issue Available**: @engcom-Charlie, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by aleromano89** [comment]
@magento-engcom-team the issue also happens on 2.4-develop

**Comment by VladimirZaets** [comment]
Hi @adrian-martinez-interactiv4. Thank you for your report. 
The issue has been fixed in magento/magento2#26264 by @aleromano89 in 2.4-develop branch
Related commit(s):
 - [95fc06f456d933aaaf8ca2b039946dc02d48483d](https://github.com/magento/magento2/commit/95fc06f456d933aaaf8ca2b039946dc02d48483d)
 - [b015b65baa6d54dbffd85f11e37d62ca1a28d5da](https://github.com/magento/magento2/commit/b015b65baa6d54dbffd85f11e37d62ca1a28d5da)
 - [05ecfda9c34698c670ff8fc302632e52fe82f9de](https://github.com/magento/magento2/commit/05ecfda9c34698c670ff8fc302632e52fe82f9de)

The fix will be available with the upcoming 2.4.0 release.
